### ISSUE TITLE
memory leak caused by bloc close method

### ISSUE THREAD

**Issue by fayduan** [felangel/bloc#2186]
**Describe the bug**
Use dart vm_service to check the memory, and find that when the widget is closed, the bloc object is still held by 
 `_Future._awaiter`

**To Reproduce**
1. Run `main.dart` (at end of this issue)
2. Click 'Click Me'
3. Click 'Back' on the top-left
4. Check bloc object retaining path 

**Expected behavior**
Expect the bloc is collected by gc, and not no longer held by other object.

**Logs**

- Flutter version 1.22.6
- Dart version 2.10.5
- dependency:
```yaml
dependencies:
  flutter:
    sdk: flutter
  flutter_bloc: ^6.1.2
  vm_service: ^5.5.0
```
bloc object retaining path get from vm_service as blow:
```
#0      TestBloc (package:flutter_demo/main.dart:64:1)
#1      Context(objects/216)
#2      _Closure.async_op (package:bloc/src/bloc.dart:212:16)
#3      Bloc.close (package:bloc/src/bloc.dart:210:3)
#4      Bloc (package:bloc/src/bloc.dart:18:1)
#5      _Future._awaiter (dart:async-patch/async_patch.dart:296:13)
```
I try to modify `bloc.dart` to solve the problem, when I remove the `await` for `_transitionSubscription?.cancel();` at `bloc.dart:212`, this memory leak no longer exists. 
I guess that await for subscription's cancel is the reason of memory leak.
Although `Bloc.close` method is async, is it possible that there is some bug in `InheritedProvider.dispose`?
modify as below:
```dart
@override
@mustCallSuper
Future<void> close() async {
  await _eventController.close();
  _transitionSubscription?.cancel();    // remove await this line
  return super.close();
}
```

**Additional context**
simple demo

`main.dart`
```dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_bloc/flutter_bloc.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        navigatorObservers: [AppObserver()],
        home: Builder(
          builder: (context) => Scaffold(
            appBar: AppBar(
              title: Text("Test Bloc Leak"),
            ),
            body: Center(
              child: GestureDetector(
                child: Text(
                  'Click Me',
                  style: Theme.of(context).textTheme.headline6,
                ),
                onTap: () {
                  Navigator.of(context).push(CupertinoPageRoute<BlocPage>(builder: (BuildContext context) => BlocPage()));
                },
              ),
            ),
          ),
        ));
  }
}

class BlocPage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("bloc page"),
      ),
      body: BlocProvider(
        create: (_) => TestBloc(1),
        child: Center(
          child: InnerWidget(),
        ),
      ),
    );
  }
}

class InnerWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final bloc = BlocProvider.of<TestBloc>(context);
    // record bloc object
    return Text(
      'Second Page',
    );
  }
}

class TestBloc extends Bloc<String, int> {
  TestBloc(initialState) : super(initialState);

  @override
  Stream<int> mapEventToState(String event) async* {
    yield 1;
  }
}

class AppObserver extends NavigatorObserver {
  @override
  void didPop(Route route, Route previousRoute) {
    Future.delayed(const Duration(seconds: 1), () {
      // execute vm service gc
      // check bloc object retaining path
    });
    super.didPop(route, previousRoute);
  }

  @override
  void didPush(Route route, Route previousRoute) {
    super.didPush(route, previousRoute);
  }
}

```

**Comment by felangel** [comment]
Hi @fayduan 👋 
Thanks for opening an issue!

I was able to reproduce this and after digging I'm fairly certain it's due to the internal event stream being a broadcast stream when it shouldn't be. This was fixed in the null safe release but I must've forgotten to apply the changes to master.

Can you take a look at #2187 and verify the fix on your end as well? Thanks! 🙏

**Comment by fayduan** [comment]
Hi @felangel 
#2187 fix the issue on my end.
Thanks.

**Comment by felangel** [comment]
Fix is published in [bloc v6.1.2](https://pub.dev/packages/bloc/versions/6.1.2) 👍
